programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+    {
+        size_t length = header[1] << 16 | header[2] << 8 | header[3];
+        mbedtls_serialize_item_t *item;
+        DBG( "received push length=%zu", length );
+        if( ctx->status == MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY )
+        {
+            /* While we're out of memory, keep reading arguments but discard
+               them. */
+            while( length > sizeof( header ) )
+            {
+                if( ( ret = mbedtls_serialize_read( ctx, header, sizeof( header ) ) ) != 0 )
+                {
+                    DBG( "failed to read input with %zu bytes remaining -> dead", length );
+                    ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+                    return( ret );
+                }
+                length -= sizeof( header );
+            }
+            if( ( ret = mbedtls_serialize_read( ctx, header, length ) ) != 0 )
+            {
+                DBG( "failed to read input with %zu bytes remaining -> dead", length );
+                ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+                return( ret );
+            }
+            return( MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED );
+        }
+        item = alloc_item( length );
+        if( item == NULL )
+        {
+            DBG( "failed to allocate %zu bytes for input", length );
+            ctx->status = MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY;
+            return( MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED );
+        }
+        if( ( ret = mbedtls_serialize_read( ctx, item_buffer( item ), length ) ) != 0 )
+        {
+            DBG( "failed to read %zu-byte input -> dead", length );
+            ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+            return( ret );
+        }
+        DBG( "successfully read %zu-byte input", length );
+        item->next = ctx->stack;
+        ctx->stack = item;
+        return( 0 );
+    }
+
+    case MBEDTLS_SERIALIZE_TYPE_EXECUTE:
+    {
+        uint32_t function = header[1] << 16 | header[2] << 8 | header[3];
+        uint32_t status;
+        mbedtls_serialize_uint32_t status_item = {{NULL, 4}, {0}};
+        uint8_t *status_data = item_buffer( &status_item.meta );
+        mbedtls_serialize_item_t *outputs[1 + 16] = {&status_item.meta};
+        size_t i;
+        DBG( "executing function 0x%06x", function );
+        if( ctx->status == MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY )
+        {
+            /* Send an out-of-memory status */
+            DBG( "already out of memory" );
+            status = MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED;
+        }
+        else
+        {
+            status = mbedtls_serialize_perform( ctx, function, outputs + 1 );
+        }
+        DBG( "status = 0x%08x", status );
+        status_data[0] = status >> 24 & 0xff;
+        status_data[1] = status >> 16 & 0xff;
+        status_data[2] = status >> 8 & 0xff;
+        status_data[3] = status & 0xff;
+        for( i = 0; i < sizeof( outputs ) / sizeof( *outputs ) && outputs[i] != NULL ; i++ )
```

Comment:
@hanno-arm At the moment, the encoding of the arities in the function identifier is purely a convention, it isn't used or enforced anywhere. I haven't decided yet whether it is a good convention that should become a rule, or a useless convention that should be simplified away.